python/pyspark/sql/streaming/state_api_client.py

```diff
@@ -0,0 +1,172 @@
+from pyspark.sql.types import StructType, _parse_datatype_string
+
+
+class StatefulProcessorHandleState(Enum):
+    CREATED = 1
+    INITIALIZED = 2
+    DATA_PROCESSED = 3
+    CLOSED = 4
+
+
+class StateApiClient:
+    def __init__(
+            self,
+            state_server_port: int) -> None:
+        self._client_socket = socket.socket()
+        self._client_socket.connect(("localhost", state_server_port))
+        self.sockfile = self._client_socket.makefile("rwb", int(os.environ.get("SPARK_BUFFER_SIZE",
+                                                                               65536)))
+        print(f"client is ready - connection established")
+        self.handle_state = StatefulProcessorHandleState.CREATED
+        self.utf8_deserializer = UTF8Deserializer()
+        # place holder, will remove when actual implementation is done
+        # self.setHandleState(StatefulProcessorHandleState.CLOSED)
+
+    def set_handle_state(self, state: StatefulProcessorHandleState) -> None:
+        print(f"setting handle state to: {state}")
+        proto_state = self._get_proto_state(state)
+        set_handle_state = stateMessage.SetHandleState(state=proto_state)
+        handle_call = stateMessage.StatefulProcessorCall(setHandleState=set_handle_state)
+        message = stateMessage.StateRequest(statefulProcessorCall=handle_call)
+
+        self._send_proto_message(message)
+        status = read_int(self.sockfile)
+
+        if (status == 0):
+            self.handle_state = state
+        print(f"setHandleState status= {status}")
+
+    def get_value_state(self, state_name: str, schema: Union[StructType, str]) -> None:
+        if isinstance(schema, str):
+            schema = cast(StructType, _parse_datatype_string(schema))
+
+        print(f"initializing value state: {state_name}")
+
+        state_call_command = stateMessage.StateCallCommand()
+        state_call_command.stateName = state_name
+        state_call_command.schema = schema.json()
+        call = stateMessage.StatefulProcessorCall(getValueState=state_call_command)
+
+        message = stateMessage.StateRequest(statefulProcessorCall=call)
+
+        self._send_proto_message(message)
+        status = read_int(self.sockfile)
+        print(f"getValueState status= {status}")
+
+    def value_state_exists(self, state_name: str) -> bool:
+        print(f"checking value state exists: {state_name}")
+        exists_call = stateMessage.Exists(stateName=state_name)
+        value_state_call = stateMessage.ValueStateCall(exists=exists_call)
+        state_variable_request = stateMessage.StateVariableRequest(valueStateCall=value_state_call)
+        message = stateMessage.StateRequest(stateVariableRequest=state_variable_request)
+
+        self._send_proto_message(message)
+        status = read_int(self.sockfile)
+        print(f"valueStateExists status= {status}")
+        if (status == 0):
+            return True
+        else:
+            return False
+
```

Comment:
Lets add explicit error handling here. And lets make the response type a protobuf message which can encapsulate non-existence of state.